kube/src/config/mod.rs

```diff
@@ -14,10 +14,60 @@ use crate::{Error, Result};
 pub use file_loader::KubeConfigOptions;
 use file_loader::{ConfigLoader, Der};
 
+use chrono::{DateTime, Utc};
 use reqwest::header::{self, HeaderMap};
+use tokio::sync::Mutex;
 
+use std::sync::Arc;
 use std::time::Duration;
 
+#[derive(Debug, Clone)]
+pub(crate) enum Authentication {
+    None,
+    Basic(String),
+    Token(String),
+    RefreshableToken(Arc<Mutex<(String, DateTime<Utc>)>>, ConfigLoader),
+}
+
+impl Authentication {
+    async fn to_header(&self) -> Result<Option<header::HeaderValue>> {
+        match self {
+            Self::None => Ok(None),
+            Self::Basic(value) => {
+                Ok(Some(header::HeaderValue::from_str(value).map_err(|e| {
+                    Error::Kubeconfig(format!("Invalid basic auth: {}", e))
+                })?))
+            }
+            Self::Token(value) => {
+                Ok(Some(header::HeaderValue::from_str(value).map_err(|e| {
+                    Error::Kubeconfig(format!("Invalid bearer token: {}", e))
+                })?))
+            }
+            Self::RefreshableToken(data, loader) => {
+                let mut locked_data = data.lock().await;
+                // Add some wiggle room onto the current timestamp so we don't get any race
+                // conditions where the token expires while we are refreshing
+                if chrono::Utc::now() + chrono::Duration::seconds(60) >= locked_data.1 {
+                    if let Authentication::RefreshableToken(d, _) = load_auth_header(loader)? {
+                        let (new_token, new_expire) = Arc::try_unwrap(d)
```

Comment:
`d` in theory is a new token straight out of the library right? This unpacking of Arc<Mutex should never fail because we are making a new one?

Reply:
Yeah, it should never fail because there should only be one reference, but there is only a `try_unwrap` method, so I handled the error in the super odd case that this somehow fails

Reply:
To avoid the confusion this error-that-can't-happen produces, perhaps `load_auth_header` should produce something that is already wrapped, then the two other functions can wrap them?

Reply:
I would just `expect`. This is a programming error on the part of the `kube` crate if it fails. We don't want that to propagate to the user.

Reply:
@clux We can't unwrap it or we'd have to take it out of the enum and use something like `Either`. Can do that if you prefer.

I am going to switch it to @rylev's suggestion for now as it should panic as it is a programming error

Reply:
Yeah, that's probably better.